sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbGridFSIO.java

```diff
@@ -73,132 +79,167 @@
  * <p>The source also accepts an optional configuration: {@code withQueryFilter()} allows you to
  * define a JSON filter to get subset of files in the database.</p>
  *
- * <p>There is also an optional {@code ParseCallback} that can be specified that can be used to
+ * <p>There is also an optional {@code Parser} that can be specified that can be used to
  * parse the InputStream into objects usable with Beam.  By default, MongoDbGridFSIO will parse
  * into Strings, splitting on line breaks and using the uploadDate of the file as the timestamp.
+ * When using a parser that outputs with custom timestamps, you may also need to specify
+ * the allowedTimestampSkew option.
  */
 public class MongoDbGridFSIO {
 
   /**
-   * Function for parsing the GridFSDBFile into objects for the PCollection.
+   * Callback for the parser to use to submit data.
+   */
+  public interface ParserCallback<T> extends Serializable {
+    public void output(T output);
+
+    public void output(T output, @Nullable Instant timestamp);
```

Comment:
Actually, there is a forth as well that I've seen used:  the GridFSDBFile can have additional metadata attached to it when saved to the grid which can include information about where/when the file came from.   For example, a REST service that accepts a file of some sort and sticks it in the grid can result in the ObjectIDs timestamp and uploadDate being the time the file is stuck in the grid, but metadata records the original creation timestamp that the user passes in.